jwst/assign_mtwcs/moving_target_wcs.py

```diff
@@ -16,57 +16,63 @@
+    # loop over only science exposures in the ModelLibrary
+    ind = input_models.indices_for_exptype("science")
+    mt_ra = np.empty(len(ind))
+    mt_dec = np.empty(len(ind))
+    with input_models:
+        for i in ind:
+            model = input_models.borrow(i)
+            mt_ra[i] = model.meta.wcsinfo.mt_ra
+            mt_dec[i] = model.meta.wcsinfo.mt_dec
+            input_models.shelve(model, i, modify=False)
 
     # Compute the mean MT RA/Dec over all exposures
     if None in mt_ra or None in mt_dec:
         log.warning("One or more MT RA/Dec values missing in input images")
         log.warning("Step will be skipped, resulting in target misalignment")
-        for model in sci_models:
-            model.meta.cal_step.assign_mtwcs = 'SKIPPED'
-        return input_model
-    else:
-        mt_avra = mt_ra.mean()
-        mt_avdec = mt_dec.mean()
-
-    for model in sci_models:
-        model.meta.wcsinfo.mt_avra = mt_avra
-        model.meta.wcsinfo.mt_avdec = mt_avdec
-        if isinstance(model, datamodels.MultiSlitModel):
-            for ind, slit in enumerate(model.slits):
-                new_wcs = add_mt_frame(slit.meta.wcs,
-                                       mt_avra, mt_avdec,
-                                       slit.meta.wcsinfo.mt_ra, slit.meta.wcsinfo.mt_dec)
-                del model.slits[ind].meta.wcs
-                model.slits[ind].meta.wcs = new_wcs
-        else:
-
-            new_wcs = add_mt_frame(model.meta.wcs, mt_avra, mt_avdec,
-                                   model.meta.wcsinfo.mt_ra, model.meta.wcsinfo.mt_dec)
-            del model.meta.wcs
-            model.meta.wcs = new_wcs
-
-        model.meta.cal_step.assign_mtwcs = 'COMPLETE'
-
-    return input_model
+        record_step_status(input_models, "assign_mtwcs", False)
+        return input_models
+
+    mt_avra = mt_ra.mean()
+    mt_avdec = mt_dec.mean()
+
+    with input_models:
+        for i in ind:
+            model = input_models.borrow(i)
+            model.meta.wcsinfo.mt_avra = mt_avra
+            model.meta.wcsinfo.mt_avdec = mt_avdec
+            if isinstance(model, datamodels.MultiSlitModel):
+                for ind, slit in enumerate(model.slits):
+                    new_wcs = add_mt_frame(slit.meta.wcs,
+                                        mt_avra, mt_avdec,
+                                        slit.meta.wcsinfo.mt_ra, slit.meta.wcsinfo.mt_dec)
+                    del model.slits[ind].meta.wcs
+                    model.slits[ind].meta.wcs = new_wcs
+            else:
+
+                new_wcs = add_mt_frame(model.meta.wcs, mt_avra, mt_avdec,
+                                    model.meta.wcsinfo.mt_ra, model.meta.wcsinfo.mt_dec)
+                del model.meta.wcs
+                model.meta.wcs = new_wcs
+            input_models.shelve(model, i, modify=True)
+            
+    record_step_status(input_models, "assign_mtwcs", True)
```

Comment:
```suggestion
                model.meta.wcs = new_wcs
            record_step_status(model, "assign_mtwcs", True)
            input_models.shelve(model, i, modify=True)
```
I believe the previous code only set the cal_step for the science models (the assignment occurs in a `for model in sci_models` loop). This suggestion moves the assignment back within the loop.

Reply:
I don't think it matters either way whether the non-science models get their status updated, but your way does avoid another load...save operation if the library is on disk